fdbclient/ServerKnobs.cpp

```diff
@@ -33,7 +33,7 @@ void ServerKnobs::initialize(Randomize randomize, ClientKnobs* clientKnobs, IsSi
 	init( VERSIONS_PER_SECOND,                                   1e6 );
 	init( MAX_VERSIONS_IN_FLIGHT,                100 * VERSIONS_PER_SECOND );
 	init( MAX_VERSIONS_IN_FLIGHT_FORCED,         6e5 * VERSIONS_PER_SECOND ); //one week of versions
-	init( MAX_READ_TRANSACTION_LIFE_VERSIONS,      5 * VERSIONS_PER_SECOND ); if (randomize && BUGGIFY) MAX_READ_TRANSACTION_LIFE_VERSIONS = VERSIONS_PER_SECOND; else if (randomize && BUGGIFY) MAX_READ_TRANSACTION_LIFE_VERSIONS = std::max<int>(1, 0.1 * VERSIONS_PER_SECOND); else if( randomize && BUGGIFY ) MAX_READ_TRANSACTION_LIFE_VERSIONS = 10 * VERSIONS_PER_SECOND;
+	init( MAX_READ_TRANSACTION_LIFE_VERSIONS,      5 * VERSIONS_PER_SECOND ); if (randomize && BUGGIFY) MAX_READ_TRANSACTION_LIFE_VERSIONS = VERSIONS_PER_SECOND; else if (randomize && BUGGIFY && !ENABLE_VERSION_VECTOR) MAX_READ_TRANSACTION_LIFE_VERSIONS = std::max<int>(1, 0.1 * VERSIONS_PER_SECOND); else if( randomize && BUGGIFY ) MAX_READ_TRANSACTION_LIFE_VERSIONS = 10 * VERSIONS_PER_SECOND;
```

Comment:
The method "initKnob()" hasn't been invoked on ENABLE_VERSION_VECTOR yet, so how is this working?

Reply:
Below is the test I did to verify this. 

```
init( MAX_READ_TRANSACTION_LIFE_VERSIONS,      1e5 );  if (!ENABLE_VERSION_VECTOR) MAX_READ_TRANSACTION_LIFE_VERSIONS=5*1e6;
init( ENABLE_VERSION_VECTOR,                                true );
init( ENABLE_VERSION_VECTOR_TLOG_UNICAST,                   true );

./build_output/bin/fdbserver -r simulation --logsize=4096MiB -f ./src/foundationdb/tests/fast/SwizzledRollba^CSideband.toml -b on -s 971188171;
cmr |grep -i MAX_READ_TRANSACTION_LIFE_VERSIONS
0.000000" Knob ID=0000000000000000 Name="max_read_transaction_life_versions" Value="100000" Atomic="0" ThreadID="5987166842268595438"  


./build_output/bin/fdbserver -r simulation --logsize=4096MiB --knob_enable_version_vector=0 --knob_enable_version_vector_tlog_unicast=0 -f ./src/foundationdb/tests/fast/SwizzledRollba^CSideband.toml -b on -s 971188171;
cmr |s|grep -i MAX_READ_TRANSACTION_LIFE_VERSIONS
0.000000" Knob ID=0000000000000000 Name="max_read_transaction_life_versions" Value="5000000" Atomic="0" ThreadID="5987166842268595438"  
```

Reply:
I am not sure how specifying "--knob_" argument to "fdbserver" actually works. Maybe you can run the test, without specifying that argument, over your change and check the parameter value? If ENABLE_VERSION_VECTOR is not initialized by the time it is used, which it appears is the case here, then your change might not be effective (and, worse, we might be introducing non-determinism here). But I will let Jingyu comment on whether this works correctly or not.

Reply:
Without --knob, the test also gives expected results. I agree from the standpoint of code inspection it would appear code after the init() macro would be executed before later init()s are run, however that is *not* the behavior I see in the test. To remove ambiguity I can update the PR to reorder the knobs anyway.